File: backend/src/validation.rs
use regex::Regex;
use std::collections::HashSet;

/// Maximum length for names and descriptions
const MAX_NAME_LENGTH: usize = 255;
const MAX_DESCRIPTION_LENGTH: usize = 1000;
const MAX_COMMAND_LENGTH: usize = 1000;
const MAX_IMAGE_LENGTH: usize = 255;

/// Sanitize and validate a name field
pub fn sanitize_name(name: &str) -> Result<String, String> {
    if name.is_empty() {
        return Err("Name cannot be empty".to_string());
    }
    
    if name.len() > MAX_NAME_LENGTH {
        return Err(format!("Name too long: {} characters (max {})", name.len(), MAX_NAME_LENGTH));
    }
    
    // Allow alphanumeric, spaces, hyphens, underscores, and periods
    let allowed_chars_regex = Regex::new(r"^[a-zA-Z0-9\s._-]+$").unwrap();
    if !allowed_chars_regex.is_match(name) {
        return Err("Name contains invalid characters. Only alphanumeric characters, spaces, hyphens, underscores, and periods are allowed".to_string());
    }
    
    // Trim and normalize whitespace
    let sanitized = name.trim().to_string();
    
    // Prevent names that are just whitespace
    if sanitized.is_empty() {
        return Err("Name cannot be empty after trimming".to_string());
    }
    
    Ok(sanitized)
}

/// Sanitize and validate a description field
pub fn sanitize_description(description: &str) -> Result<String, String> {
    if description.len() > MAX_DESCRIPTION_LENGTH {
        return Err(format!("Description too long: {} characters (max {})", description.len(), MAX_DESCRIPTION_LENGTH));
    }
    
    // Remove potential HTML/XML tags and harmful characters
    let sanitized = description
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('&', "&amp;")
        .replace('"', "&quot;")
        .replace('\'', "&#x27;")
        .replace('/', "&#x2F;");
    
    Ok(sanitized.trim().to_string())
}

/// Validate Docker image name format
pub fn validate_docker_image(image: &str) -> Result<String, String> {
    if image.is_empty() {
        return Err("Docker image cannot be empty".to_string());
    }
    
    if image.len() > MAX_IMAGE_LENGTH {
        return Err(format!("Docker image name too long: {} characters (max {})", image.len(), MAX_IMAGE_LENGTH));
    }
    
    // Docker image name regex: [registry/]namespace/repository[:tag]
    // This is a simplified version - real docker image validation is complex
    let image_regex = Regex::new(r"^[a-zA-Z0-9._-]+(?:/[a-zA-Z0-9._-]+)*(?::[a-zA-Z0-9._-]+)?$").unwrap();
    if !image_regex.is_match(image) {
        return Err("Invalid Docker image format. Must be in format: [registry/]namespace/repository[:tag]".to_string());
    }
    
    // Prevent some obviously malicious patterns
    let dangerous_patterns = ["../", "..\\", "$(", "`", ";", "&", "|", "&&", "||"];
    for pattern in &dangerous_patterns {
        if image.contains(pattern) {
            return Err("Docker image contains potentially dangerous characters".to_string());
        }
    }
    
    Ok(image.trim().to_string())
}

/// Sanitize and validate command array
pub fn sanitize_commands(commands: &[String]) -> Result<Vec<String>, String> {
    if commands.is_empty() {
        return Err("Commands cannot be empty".to_string());
    }
    
    if commands.len() > 100 {
        return Err("Too many commands (max 100)".to_string());
    }
    
    let mut sanitized_commands = Vec::new();
    
    for command in commands {
        if command.len() > MAX_COMMAND_LENGTH {
            return Err(format!("Command too long: {} characters (max {})", command.len(), MAX_COMMAND_LENGTH));
        }
        
        // Trim whitespace
        let trimmed = command.trim();
        if trimmed.is_empty() {
            continue; // Skip empty commands
        }
        
        // Check for dangerous shell patterns
        let dangerous_patterns = [
            "$(", "`", ";", "&", "|", "&&", "||", ">", "<", ">>", "<<",
            "rm -rf", "dd if=", ":(){ :|:& };:", "chmod -R", "chown -R"
        ];
        
        for pattern in &dangerous_patterns {
            if trimmed.contains(pattern) {
                return Err(format!("Command contains potentially dangerous pattern: {}", pattern));
            }
        }
        
        sanitized_commands.push(trimmed.to_string());
    }
    
    if sanitized_commands.is_empty() {
        return Err("No valid commands found after sanitization".to_string());
    }
    
    Ok(sanitized_commands)
}

/// Validate and sanitize labels
pub fn sanitize_labels(labels: &[String]) -> Result<Vec<String>, String> {
    if labels.len() > 50 {
        return Err("Too many labels (max 50)".to_string());
    }
    
    let mut sanitized_labels = Vec::new();
    let mut seen_labels = HashSet::new();
    
    for label in labels {
        if label.len() > 100 {
            return Err(format!("Label too long: {} characters (max 100)", label.len()));
        }
        
        // Allow alphanumeric, hyphens, underscores, and periods
        let label_regex = Regex::new(r"^[a-zA-Z0-9._-]+$").unwrap();
        if !label_regex.is_match(label) {
            return Err("Label contains invalid characters. Only alphanumeric characters, hyphens, underscores, and periods are allowed".to_string());
        }
        
        let trimmed = label.trim().to_lowercase();
        if trimmed.is_empty() {
            continue; // Skip empty labels
        }
        
        // Prevent duplicates
        if seen_labels.contains(&trimmed) {
            continue;
        }
        
        seen_labels.insert(trimmed.clone());
        sanitized_labels.push(trimmed);
    }
    
    Ok(sanitized_labels)
}

/// Validate execution mode for test suites
pub fn validate_execution_mode(mode: &str) -> Result<String, String> {
    match mode.to_lowercase().as_str() {
        "sequential" | "parallel" => Ok(mode.to_lowercase()),
        _ => Err("Execution mode must be 'sequential' or 'parallel'".to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sanitize_name_valid() {
        assert_eq!(sanitize_name("Valid Name").unwrap(), "Valid Name");
        assert_eq!(sanitize_name("test-name_123").unwrap(), "test-name_123");
        assert_eq!(sanitize_name("  spaced  ").unwrap(), "spaced");
    }

    #[test]
    fn test_sanitize_name_invalid() {
        assert!(sanitize_name("").is_err());
        assert!(sanitize_name("   ").is_err());
        assert!(sanitize_name("name@with$pecial").is_err());
        assert!(sanitize_name(&"a".repeat(300)).is_err());
    }

    #[test]
    fn test_sanitize_description_valid() {
        assert_eq!(sanitize_description("Valid description").unwrap(), "Valid description");
        assert_eq!(sanitize_description("").unwrap(), "");
    }

    #[test]
    fn test_sanitize_description_escapes_html() {
        let desc = "<script>alert('xss')</script>";
        let result = sanitize_description(desc).unwrap();
        assert!(!result.contains("<script>"));
        assert!(result.contains("&amp;lt;script&amp;gt;"));
    }

    #[test]
    fn test_validate_docker_image_valid() {
        assert_eq!(validate_docker_image("nginx:latest").unwrap(), "nginx:latest");
        assert_eq!(validate_docker_image("registry.com/user/repo:tag").unwrap(), "registry.com/user/repo:tag");
        assert_eq!(validate_docker_image("ubuntu").unwrap(), "ubuntu");
    }

    #[test]
    fn test_validate_docker_image_invalid() {
        assert!(validate_docker_image("").is_err());
        assert!(validate_docker_image("image$(whoami)").is_err());
        assert!(validate_docker_image("image;rm -rf /").is_err());
        assert!(validate_docker_image("../malicious").is_err());
    }

    #[test]
    fn test_sanitize_commands_valid() {
        let commands = vec!["echo".to_string(), "hello".to_string()];
        let result = sanitize_commands(&commands).unwrap();
        assert_eq!(result, vec!["echo", "hello"]);
    }

    #[test]
    fn test_sanitize_commands_invalid() {
        assert!(sanitize_commands(&[]).is_err());
        assert!(sanitize_commands(&["rm -rf /".to_string()]).is_err());
        assert!(sanitize_commands(&["echo $(whoami)".to_string()]).is_err());
        assert!(sanitize_commands(&["cmd;malicious".to_string()]).is_err());
    }

    #[test]
    fn test_sanitize_labels_valid() {
        let labels = vec!["test".to_string(), "production".to_string()];
        let result = sanitize_labels(&labels).unwrap();
        assert_eq!(result, vec!["test", "production"]);
    }

    #[test]
    fn test_sanitize_labels_removes_duplicates() {
        let labels = vec!["test".to_string(), "TEST".to_string(), "test".to_string()];
        let result = sanitize_labels(&labels).unwrap();
        assert_eq!(result, vec!["test"]);
    }

    #[test]
    fn test_validate_execution_mode() {
        assert_eq!(validate_execution_mode("sequential").unwrap(), "sequential");
        assert_eq!(validate_execution_mode("PARALLEL").unwrap(), "parallel");
        assert!(validate_execution_mode("invalid").is_err());
    }
}

File: apps/oss/utils/validation.ts
/**
 * Frontend validation utilities for sanitizing user inputs
 */

// Maximum length constraints
const MAX_NAME_LENGTH = 255
const MAX_DESCRIPTION_LENGTH = 1000
const MAX_COMMAND_LENGTH = 1000
const MAX_IMAGE_LENGTH = 255

export interface ValidationResult {
  isValid: boolean
  error?: string
  sanitizedValue?: string
}

export interface ValidationArrayResult {
  isValid: boolean
  error?: string
  sanitizedValue?: string[]
}

/**
 * Sanitize and validate a name field
 */
export function sanitizeName(name: string): ValidationResult {
  if (!name || name.trim().length === 0) {
    return { isValid: false, error: "Name cannot be empty" }
  }

  if (name.length > MAX_NAME_LENGTH) {
    return {
      isValid: false,
      error: `Name too long: ${name.length} characters (max ${MAX_NAME_LENGTH})`,
    }
  }

  // Allow alphanumeric, spaces, hyphens, underscores, and periods
  const allowedCharsRegex = /^[a-zA-Z0-9\s._-]+$/
  if (!allowedCharsRegex.test(name)) {
    return {
      isValid: false,
      error:
        "Name contains invalid characters. Only alphanumeric characters, spaces, hyphens, underscores, and periods are allowed",
    }
  }

  // Trim and normalize whitespace
  const sanitized = name.trim()

  return { isValid: true, sanitizedValue: sanitized }
}

/**
 * Sanitize and validate a description field
 */
export function sanitizeDescription(description: string): ValidationResult {
  if (description.length > MAX_DESCRIPTION_LENGTH) {
    return {
      isValid: false,
      error: `Description too long: ${description.length} characters (max ${MAX_DESCRIPTION_LENGTH})`,
    }
  }

  // Remove potential HTML/XML tags and harmful characters
  const sanitized = description
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/&/g, "&amp;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#x27;")
    .replace(/\//g, "&#x2F;")

  return { isValid: true, sanitizedValue: sanitized.trim() }
}

/**
 * Validate Docker image name format
 */
export function validateDockerImage(image: string): ValidationResult {
  if (!image || image.trim().length === 0) {
    return { isValid: false, error: "Docker image cannot be empty" }
  }

  if (image.length > MAX_IMAGE_LENGTH) {
    return {
      isValid: false,
      error: `Docker image name too long: ${image.length} characters (max ${MAX_IMAGE_LENGTH})`,
    }
  }

  // Prevent some obviously malicious patterns first
  const dangerousPatterns = ["../", "..\\", "$(", "`", ";", "&", "|", "&&", "||"]
  for (const pattern of dangerousPatterns) {
    if (image.includes(pattern)) {
      return {
        isValid: false,
        error: "Docker image contains potentially dangerous characters",
      }
    }
  }

  // Docker image name validation - simplified but secure
  // Allow alphanumeric, dots, hyphens, underscores, slashes, and colons
  const imageRegex = /^[a-zA-Z0-9._/-]+(?::[a-zA-Z0-9._-]+)?$/
  if (!imageRegex.test(image)) {
    return {
      isValid: false,
      error:
        "Invalid Docker image format. Must be in format: [registry/]namespace/repository[:tag]",
    }
  }

  return { isValid: true, sanitizedValue: image.trim() }
}

/**
 * Sanitize and validate command array
 */
export function sanitizeCommands(commands: string[]): ValidationArrayResult {
  if (!commands || commands.length === 0) {
    return { isValid: false, error: "Commands cannot be empty" }
  }

  if (commands.length > 100) {
    return { isValid: false, error: "Too many commands (max 100)" }
  }

  const sanitizedCommands: string[] = []

  for (const command of commands) {
    if (command.length > MAX_COMMAND_LENGTH) {
      return {
        isValid: false,
        error: `Command too long: ${command.length} characters (max ${MAX_COMMAND_LENGTH})`,
      }
    }

    // Trim whitespace
    const trimmed = command.trim()
    if (trimmed.length === 0) {
      continue // Skip empty commands
    }

    // Check for dangerous shell patterns
    const dangerousPatterns = [
      "$(",
      "`",
      ";",
      "&",
      "|",
      "&&",
      "||",
      ">",
      "<",
      ">>",
      "<<",
      "rm -rf",
      "dd if=",
      ":(){ :|:& };:",
      "chmod -R",
      "chown -R",
      "cat /etc/",
      "cat ~/",
      "cat ~/.ssh/",
      "grep -r",
      "find /",
      "history",
      "passwd",
      "sudo",
      "su ",
      "su -",
      "crontab",
      "systemctl",
      "service ",
      "pkill",
      "killall",
      "mount",
      "umount",
      "iptables",
      "usermod",
      "useradd",
      "userdel",
      "/etc/passwd",
      "/etc/shadow",
      "/etc/hosts",
      "/dev/",
      "/proc/",
      "/sys/",
      "~/.ssh/",
      "~/.bash_history",
      "curl ",
      "wget ",
      "nc ",
      "netcat ",
      "telnet ",
      "ssh ",
      "scp ",
      "rsync ",
      "tar ",
      "zip ",
      "unzip ",
      "gzip ",
      "gunzip ",
      "python -c",
      "node -e",
      "ruby -e",
      "php -e",
      "perl -e",
      "bash -c",
      "sh -c",
      "eval ",
      "exec ",
      "system(",
      "os.system",
      "child_process",
      "require(",
      "import os",
    ]

    for (const pattern of dangerousPatterns) {
      if (trimmed.includes(pattern)) {
        return {
          isValid: false,
          error: `Command contains potentially dangerous pattern: ${pattern}`,
        }
      }
    }

    sanitizedCommands.push(trimmed)
  }

  if (sanitizedCommands.length === 0) {
    return { isValid: false, error: "No valid commands found after sanitization" }
  }

  return { isValid: true, sanitizedValue: sanitizedCommands }
}

/**
 * Validate and sanitize labels
 */
export function sanitizeLabels(labels: string[]): ValidationArrayResult {
  if (labels.length > 50) {
    return { isValid: false, error: "Too many labels (max 50)" }
  }

  const sanitizedLabels: string[] = []
  const seenLabels = new Set<string>()

  for (const label of labels) {
    if (label.length > 100) {
      return {
        isValid: false,
        error: `Label too long: ${label.length} characters (max 100)`,
      }
    }

    // Allow alphanumeric, hyphens, underscores, and periods
    const labelRegex = /^[a-zA-Z0-9._-]+$/
    if (!labelRegex.test(label)) {
      return {
        isValid: false,
        error:
          "Label contains invalid characters. Only alphanumeric characters, hyphens, underscores, and periods are allowed",
      }
    }

    const trimmed = label.trim().toLowerCase()
    if (trimmed.length === 0) {
      continue // Skip empty labels
    }

    // Prevent duplicates
    if (seenLabels.has(trimmed)) {
      continue
    }

    seenLabels.add(trimmed)
    sanitizedLabels.push(trimmed)
  }

  return { isValid: true, sanitizedValue: sanitizedLabels }
}

/**
 * Validate execution mode for test suites
 */
export function validateExecutionMode(mode: string): ValidationResult {
  const normalizedMode = mode.toLowerCase()
  if (normalizedMode === "sequential" || normalizedMode === "parallel") {
    return { isValid: true, sanitizedValue: normalizedMode }
  }
  return {
    isValid: false,
    error: "Execution mode must be 'sequential' or 'parallel'",
  }
}

/**
 * Validate form data for test definition
 */
export function validateTestDefinition(data: {
  name: string
  description?: string
  image: string
  commands: string[]
}): { isValid: boolean; errors: Record<string, string> } {
  const errors: Record<string, string> = {}

  const nameResult = sanitizeName(data.name)
  if (!nameResult.isValid) {
    errors.name = nameResult.error!
  }

  if (data.description) {
    const descResult = sanitizeDescription(data.description)
    if (!descResult.isValid) {
      errors.description = descResult.error!
    }
  }

  const imageResult = validateDockerImage(data.image)
  if (!imageResult.isValid) {
    errors.image = imageResult.error!
  }

  const commandsResult = sanitizeCommands(data.commands)
  if (!commandsResult.isValid) {
    errors.commands = commandsResult.error!
  }

  return { isValid: Object.keys(errors).length === 0, errors }
}

/**
 * Validate form data for test executor
 */
export function validateTestExecutor(data: {
  name: string
  description?: string
  image: string
  default_command: string
}): { isValid: boolean; errors: Record<string, string> } {
  const errors: Record<string, string> = {}

  const nameResult = sanitizeName(data.name)
  if (!nameResult.isValid) {
    errors.name = nameResult.error!
  }

  if (data.description) {
    const descResult = sanitizeDescription(data.description)
    if (!descResult.isValid) {
      errors.description = descResult.error!
    }
  }

  const imageResult = validateDockerImage(data.image)
  if (!imageResult.isValid) {
    errors.image = imageResult.error!
  }

  const commandsResult = sanitizeCommands([data.default_command])
  if (!commandsResult.isValid) {
    errors.default_command = commandsResult.error!
  }

  return { isValid: Object.keys(errors).length === 0, errors }
}

/**
 * Validate form data for test suite
 */
export function validateSuite(data: {
  name: string
  description?: string
  execution_mode: string
  labels: string[]
}): { isValid: boolean; errors: Record<string, string> } {
  const errors: Record<string, string> = {}

  const nameResult = sanitizeName(data.name)
  if (!nameResult.isValid) {
    errors.name = nameResult.error!
  }

  if (data.description) {
    const descResult = sanitizeDescription(data.description)
    if (!descResult.isValid) {
      errors.description = descResult.error!
    }
  }

  const executionModeResult = validateExecutionMode(data.execution_mode)
  if (!executionModeResult.isValid) {
    errors.execution_mode = executionModeResult.error!
  }

  const labelsResult = sanitizeLabels(data.labels)
  if (!labelsResult.isValid) {
    errors.labels = labelsResult.error!
  }

  return { isValid: Object.keys(errors).length === 0, errors }
}
